set.mm

```diff
@@ -645047,6 +645080,263 @@ fixed reference functional determined by this vector (corresponding to
       UWHUUDRYTUWHUUTEUYLUYKYTUXOUUEYSYKYOYRTUJYDEVHSAUUFRCDUUG $.
   $}
 
+$(
+=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
+  Sticks and stones
+=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
+$)
+
+  ${
+    $d A f $.  $d I j x y $.  $d I j z $.  $d K f x y $.  $d K j x y $.
+    $d K j z $.  $d N a $.  $d N f $.  $d X f x y $.  $d X j x y $.
+    $d X j z $.  $d Y f x y $.  $d Y j x y $.  $d Y j z $.  $d a ph $.
+    $d f ph $.  $d j ph $.
+    sticksstones1.1 $e |- ( ph -> N e. NN0 ) $.
+    sticksstones1.2 $e |- ( ph -> K e. NN0 ) $.
+    sticksstones1.3 $e |- A = { f | ( f : ( 1 ... K ) --> ( 1 ... N ) /\ A. x
+    e. ( 1 ... K ) A. y e. ( 1 ... K )
+    ( x < y -> ( f ` x ) < ( f ` y ) ) ) } $.
+    sticksstones1.4 $e |- ( ph -> X e. A ) $.
+    sticksstones1.5 $e |- ( ph -> Y e. A ) $.
+    sticksstones1.6 $e |- ( ph -> X =/= Y ) $.
+    sticksstones1.7 $e |- I = inf ( { z e. ( 1 ... K ) |
+     ( X ` z ) =/= ( Y ` z ) } , RR , < ) $.
+    $( Different strictly monotone functions have different ranges.
+       (Contributed by metakunt, 27-Sep-2024.) $)
+    sticksstones1 $p  |- ( ph -> ran X =/= ran Y ) $=
+      ( clt wcel va vj cfv wbr wo crn wne cv c1 cfz co crab cr cinf wceq a1i c0
+      cfn wss w3a syl2anc wral wn nne bitri wa wfn wb wf wi feq1 breq12d imbi2d
+      fveq1 2ralbidv anbi12d adantl ralrimiva rspcdva simpld ffnd mpdan syldbl2
+      adantr ex imp cn fz1ssnn sstrd sseldd mpbird fveq2 neeq12d syl mpbid nfcv
+      nfv elfznn nnre ffvelrnd wnel wfun cdm ffund fdmd eleqtrrd 3ad2ant3 nnred
+      fvelrn lttri4d simp3 sseli simprd simpl3 breq1 breq1d breq2 breq2d rspc2v
+      3ad2ant1 imbi12d mpd cle sylib 3expa 3adantl2 ltned necomd neeq1d 3jaodan
+      simpl2 lttrd neneqd wrex ralnex nnel fvelrnb bitrd con1bid elnelne1 fzfid
+      wor ltso ssrab2 ssfi rabeq0 ralbii cab wal abeq2 spi biimpi eqfnfv bicomd
```

Comment:
Why is ~abeq used here? It's usage is discouraged. Why not use ~abeqw instead?

Reply:
I needed a theorem to unfold the class abstraction. Not knowing which theorems to use I used the mmj2 search feature and searched on the website. This was surprisingly the only theorem I've found in mmj2.

I've conjectured that this is the only theorem that unfolds the definition and by the time I did the merge request I've figured that the theorem may not be that straightforward to change. Then I've looked at the axioms and pretty much concluded that I am using all of those axioms anyways.

Honestly I wouldn't mind refactoring it but it is very likely that my theorems are going to use the full suite of predicate calculus axioms anyways.
https://us.metamath.org/mpeuni/fvelrnb.html is used in the first theorem that uses all axioms that sticksstones1 uses and https://us.metamath.org/mpeuni/hashf1rn.html in sticksstones2.

So I don't even get any smaller axiom footprint in the final theorem.

To be fair most of my workflow comes from the point that MMJ2 autofills. I am not checking discouraged or mathbox theorems before opening a pr, so by the time it hits it is usually too late.

I guess for the future I will use elab2g, now that I know that alternatives exist to unfold the theorem.

If that's fine I'd like to keep things as is, as my workspace is currently dirty with other theorems.

Reply:
@metakunt thank you for these explanations. It looks like the problem is not using this theorem, but the tag "(New usage is discouraged.)" within the comment of ~abeq2. The usage of this theorem should only be discouraged in proofs which become dependent on more axioms using ~abeq2 instead of ~abeq2w, but this does not seem to be the case here. Unfortunately, we cannot distinguish between a general discouragement and a specific discouragement for axiom usage.

Since ~abeq2 cannot simply be replaced by ~abeq2w in a proof (probably the bound variable has to be changed for that, and this blows up the proof unnecessarily). Therefore, I propose to remove the tag "(New usage is discouraged.)" from the comment of ~abeq2.  It is absolutely correct to use this theorem in the proofs of high level theorems. @GinoGiotto, @benjub what's your opinion? Maybe other theorems of this kind are also concerned.

Reply:
> @GinoGiotto, @benjub what's your opinion? Maybe other theorems of this kind are also concerned.

Yes, I'd remove the tag. Proving theorems is already difficult enough. We don’t need to make it harder by restricting the use of axioms that are still common in the database. I think it's fine to let people prove theorems whatever way they want and optimize those proofs later.

Reply:
I agree, remove the discouragement tag from ~abeq2.  The reason to avoid ~abeq2 does exist but it is not strong enough to put the tag.

Note: a standard format for these tags is:
```
(* Blabla.  (Contributed by XX, 11-Jan-2024.)  [Brief explanation why proof modification
   is discouraged.]  (Proof modification is discouraged).  [Brief indication why new usage
   is discouraged.]  Use ~ xxx instead.  (New usage is discouraged.) $)
```

Mnemonic why tags are in that order (when both are present): you prove a result before you can use it.

Reply:
I have removed the discouragement tag.